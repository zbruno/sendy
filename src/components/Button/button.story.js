import React from 'react';
import { faSignInAlt } from '@fortawesome/free-solid-svg-icons';

import Button from './index';

export default {
  title: 'Button',
  component: Button,
  parameters: {
    jest: [],
  },
};

export const buttons = () => (
  <div className="flex flex-col items-start">
    <Button utilityClasses="mb-4">
      Primary Button
    </Button>
    <Button icon={faSignInAlt} utilityClasses="mb-4">
      Primary Icon
    </Button>
    <Button isLoading utilityClasses="mb-4">
      Loading
    </Button>
    <Button icon={faSignInAlt} isLoading utilityClasses="mb-4">
      Icon Loading
    </Button>
    <Button buttonType="secondary" utilityClasses="mb-4">
      Secondary Button
    </Button>
    <Button buttonType="secondary" icon={faSignInAlt} utilityClasses="mb-4">
      Secondary Icon
    </Button>
  </div>
);
