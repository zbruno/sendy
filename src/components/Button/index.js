import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { Loader } from '../../index';

// eslint-disable-next-line react/prefer-stateless-function
class Button extends React.Component {
  static propTypes = {
    onClick: PropTypes.func,
    isDisabled: PropTypes.bool,
    children: PropTypes.node.isRequired,
    buttonType: PropTypes.oneOf(['primary', 'secondary']),
    isLoading: PropTypes.bool,
    icon: PropTypes.shape({}),
    utilityClasses: PropTypes.string,
  };

  static defaultProps = {
    isDisabled: false,
    buttonType: 'primary',
    isLoading: false,
    onClick() {},
    icon: null,
    utilityClasses: '',
  };

  render() {
    const {
      onClick,
      isDisabled,
      children,
      buttonType,
      isLoading,
      icon,
      utilityClasses,
    } = this.props;
    const shouldAppearDisabled = isDisabled || isLoading;
    const btnClasses = cx({
      btn: true,
      [buttonType]: true,
      loading: isLoading,
      disabled: shouldAppearDisabled,
      [utilityClasses]: true,
    });
    const colorClass = cx({
      'border-white': buttonType === 'primary' && !shouldAppearDisabled,
      'border-purple-500': buttonType === 'secondary' && !shouldAppearDisabled,
      'border-gray-500': shouldAppearDisabled,
    });

    return (
      <button
        className={btnClasses}
        disabled={shouldAppearDisabled}
        type="button"
        onClick={
          !shouldAppearDisabled ? onClick : event => event.preventDefault()
        }
      >
        {icon && (
          <FontAwesomeIcon
            icon={icon}
            className="btn-icon"
            size="1x"
            fixedWidth
          />
        )}
        <span className="btn-text">
          {children}
        </span>
        <Loader
          size={6}
          color={colorClass}
          utilityClasses="btn-loader"
        />
      </button>
    );
  }
}

export default Button;
