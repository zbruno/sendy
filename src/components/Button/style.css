.btn {
  @apply relative font-bold py-3 px-8 rounded inline-flex items-center border border-transparent;

  &.primary{
    @apply text-white bg-purple-500;

    &:hover {
      @apply bg-purple-600;
    }
  }

  &.secondary {
    @apply text-purple-500 bg-transparent border-purple-500;

    &:hover {
      @apply text-purple-600 border-purple-600;
    }
  }

  &.disabled {
    @apply cursor-not-allowed bg-gray-300 text-gray-500 pointer-events-none;
  }

  & * {
    @apply transition-all;
  }

  & .btn-text {
    transform: translateX(0);
  }

  & .btn-icon {
    @apply fill-current mr-2;
    transform: translateX(0);
  }

  & .btn-loader {
    @apply absolute inset-0 left-auto mr-4 mt-2 transition-all opacity-0;
    transform: translateX(20px);
  }

  &.loading {
    & .btn-text {
      transform: translateX(-15px);
    }

    & .btn-icon {
      transform: translateX(-15px);
    }

    & .btn-loader {
      @apply opacity-100;
      transform: translateX(0);
    }
  }
}



