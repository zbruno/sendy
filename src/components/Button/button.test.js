import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { faSignInAlt } from '@fortawesome/free-solid-svg-icons';

import Button from './index';

describe('<Button />', () => {
  describe('UI snapshots', () => {
    test('renders primary with icon', () => {
      const tree = renderer
        .create(
          <Button onClick={() => { }} icon={faSignInAlt}>
            Icon Button
          </Button>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    test('renders secondary', () => {
      const tree = renderer
        .create(
          <Button onClick={() => { }} buttonType="secondary">
            Button Text
          </Button>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    test('renders loading', () => {
      const tree = renderer
        .create(
          <Button onClick={() => { }} isLoading>
            Button Text
          </Button>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    test('renders disabled', () => {
      const tree = renderer
        .create(
          <Button onClick={() => { }} isDisabled>
            Button Text
          </Button>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('onClick callback', () => {
    test('should be invoked onClick', () => {
      const spy = jest.fn();
      const button = mount(<Button onClick={spy}>Button Text</Button>);

      button.simulate('click');
      expect(spy).toHaveBeenCalled();
    });

    test('should not be invoked if disabled', () => {
      const spy = jest.fn();
      const button = mount(
        <Button isDisabled onClick={spy}>
          Button Text
        </Button>,
      );

      button.simulate('click');
      expect(spy).not.toHaveBeenCalled();
    });

    test('should not be invoked if loading', () => {
      const spy = jest.fn();
      const button = mount(
        <Button isLoading onClick={spy}>
          Button Text
        </Button>,
      );

      button.simulate('click');
      expect(spy).not.toHaveBeenCalled();
    });
  });
});
