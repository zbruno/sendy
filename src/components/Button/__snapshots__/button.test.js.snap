// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> UI snapshots renders disabled 1`] = `
<button
  className="btn primary disabled "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <span
    className="btn-text"
  >
    Button Text
  </span>
  <div
    className="inline-block h-6 w-6 btn-loader"
  >
    <div
      className="inner-loader border-gray-500"
    />
    <div
      className="inner-loader border-gray-500 animation-delay-.5s"
    />
  </div>
</button>
`;

exports[`<Button /> UI snapshots renders loading 1`] = `
<button
  className="btn primary loading disabled "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <span
    className="btn-text"
  >
    Button Text
  </span>
  <div
    className="inline-block h-6 w-6 btn-loader"
  >
    <div
      className="inner-loader border-gray-500"
    />
    <div
      className="inner-loader border-gray-500 animation-delay-.5s"
    />
  </div>
</button>
`;

exports[`<Button /> UI snapshots renders primary with icon 1`] = `
<button
  className="btn primary "
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-sign-in-alt fa-w-16 fa-fw fa-1x btn-icon"
    data-icon="sign-in-alt"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M416 448h-84c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h84c17.7 0 32-14.3 32-32V160c0-17.7-14.3-32-32-32h-84c-6.6 0-12-5.4-12-12V76c0-6.6 5.4-12 12-12h84c53 0 96 43 96 96v192c0 53-43 96-96 96zm-47-201L201 79c-15-15-41-4.5-41 17v96H24c-13.3 0-24 10.7-24 24v96c0 13.3 10.7 24 24 24h136v96c0 21.5 26 32 41 17l168-168c9.3-9.4 9.3-24.6 0-34z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <span
    className="btn-text"
  >
    Icon Button
  </span>
  <div
    className="inline-block h-6 w-6 btn-loader"
  >
    <div
      className="inner-loader border-white"
    />
    <div
      className="inner-loader border-white animation-delay-.5s"
    />
  </div>
</button>
`;

exports[`<Button /> UI snapshots renders secondary 1`] = `
<button
  className="btn secondary "
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <span
    className="btn-text"
  >
    Button Text
  </span>
  <div
    className="inline-block h-6 w-6 btn-loader"
  >
    <div
      className="inner-loader border-purple-500"
    />
    <div
      className="inner-loader border-purple-500 animation-delay-.5s"
    />
  </div>
</button>
`;
